### PR TITLE
Hparams: Use DataProvider.list_hyperparameters() to generate HparamInfos

### PR DESCRIPTION
**Pull request by bmd3k**
Use the result from a DataProvider.list_hyperparameters() call to generate HparamInfos for operations in the Hparams plugin. Generate the HparamInfo name, display_name, data_type, and domain from each Hyperparameter.

We use the same two-step strategy for generating these as is done with tensor-based hparams. The tensor-based hparams has one operation, `hparams_metadata()` for retrieving the tensor data, and then passes the result to `experiment_metadata()` to generate the HparamInfos and MetricInfos. Similarly, we add an `hparams_from_data_provider()` operation for retrieving the hyperparameter data and we pass the result to `experiment_metadata()`.

Note: We don't yet generate `MetricInfos`. We don't yet have enough information from the data provider.

Note: We don't yet merge the new HparamInfos with the tensor-based HparamInfos.

### CODE REVIEW

**Review by yatbear — CHANGES_REQUESTED** [review]

File: tensorboard/plugins/hparams/backend_context.py [hunk 1]
```diff
@@ -51,33 +52,56 @@ def __init__(self, tb_context, max_domain_discrete_len=10):
...
+        experiment_from_runs = self._compute_experiment_from_runs(
             ctx, experiment_id, hparams_run_to_tag_to_content
         )
+        if experiment_from_runs:
+            return experiment_from_runs
+
+        experiment_from_data_provider = self._experiment_from_data_provider(
+            hparams_from_data_provider
+        )
+        if experiment_from_data_provider:
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/plugins/hparams/backend_context.py]: Done.

File: tensorboard/plugins/hparams/backend_context.py [hunk 2]
```diff
@@ -273,6 +306,69 @@ def _compute_hparam_info_from_values(self, name, values):
...
         return result
 
+    def _experiment_from_data_provider(
+        self,
+        hparams_from_data_provider,
+    ):
+        """Returns an experiment protobuffer based on data provider hparams.
+
+        Args:
+          hparams_from_data_provider: The ouput from an data_provider_hparams()
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/plugins/hparams/backend_context.py]: The real typo was I used the wrong function name. :). I changed it to "an hparams_from_data_provider() call"

File: tensorboard/plugins/hparams/backend_context.py [hunk 3]
```diff
@@ -273,6 +306,69 @@ def _compute_hparam_info_from_values(self, name, values):
...
+            return None
+
+        hparam_infos = [
+            self._convert_data_provider_hparam(dp_hparam)
+            for dp_hparam in hparams_from_data_provider
+        ]
+        return api_pb2.Experiment(hparam_infos=hparam_infos)
+
+    def _convert_data_provider_hparam(self, dp_hparam):
+        """Builds am HParamInfo message from data provider Hyperparameter.
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/plugins/hparams/backend_context.py]: Done.

File: tensorboard/plugins/hparams/backend_context.py [hunk 4]
```diff
@@ -273,6 +306,69 @@ def _compute_hparam_info_from_values(self, name, values):
...
+
+        Args:
+          dp_hparams: The provider.Hyperparameter returned by the call to
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/plugins/hparams/backend_context.py]: Done.

File: tensorboard/plugins/hparams/backend_context_test.py [hunk 5]
```diff
@@ -344,31 +354,156 @@ def test_experiment_without_experiment_tag_many_distinct_values(self):
...
+              type: DATA_TYPE_FLOAT64
+              domain_interval: {
+                min_value: -10.0
+                max_value: 15
+              }
+            }
+        """
+        self.assertProtoEquals(expected_exp, actual_exp)
+
+    def test_experiment_from_data_provider_discreet_bool_hparam(self):
```
[the remark this answers is not in the document]
Comment (bmd3k) [reply tensorboard/plugins/hparams/backend_context_test.py]: Done. Fixed in three places.

File: tensorboard/plugins/hparams/backend_context.py [hunk 6]
```diff
@@ -273,6 +306,69 @@ def _compute_hparam_info_from_values(self, name, values):
...
+        hparam_info = api_pb2.HParamInfo(
+            name=dp_hparam.hyperparameter_name,
+            display_name=dp_hparam.hyperparameter_display_name,
+        )
+        if dp_hparam.domain_type == provider.HyperparameterDomainType.INTERVAL:
+            hparam_info.type = api_pb2.DATA_TYPE_FLOAT64
+            (dp_hparam_min, dp_hparam_max) = dp_hparam.domain
+            hparam_info.domain_interval.min_value = dp_hparam_min
+            hparam_info.domain_interval.max_value = dp_hparam_max
+        elif (
```
[the remark this answers is not in the document]
Comment (yatbear) [reply tensorboard/plugins/hparams/backend_context.py]: SG! Not sure if necessary - do we need to log any warning if there's an unknown type for debugging purpose here?

**Pull request merged**